lib/Tool/Frontend.php

```diff
@@ -51,25 +53,69 @@ public static function isDocumentInCurrentSite(Document $document): bool
 
     public static function getSiteForDocument(Document $document): ?Site
     {
-        $cacheKey = 'sites_full_list';
-        if (RuntimeCache::isRegistered($cacheKey)) {
-            $sites = RuntimeCache::get($cacheKey);
+        $siteIdOfDocument = self::getSiteIdForDocument($document);
+
+        if(!$siteIdOfDocument) {
+            return null;
+        }
+
+        if (RuntimeCache::isRegistered(SELF::SITE_CACHE_KEY)) {
+            $sites = RuntimeCache::get(SELF::SITE_CACHE_KEY);
         } else {
-            $sites = new Site\Listing();
-            $sites->setOrderKey('(SELECT LENGTH(`path`) FROM documents WHERE documents.id = sites.rootId) DESC', false);
-            $sites = $sites->load();
-            RuntimeCache::set($cacheKey, $sites);
+            $sites = self::getFullSiteListing();
         }
 
         foreach ($sites as $site) {
-            if (strpos($document->getRealFullPath(), $site->getRootPath() . '/') === 0 || $site->getRootDocument()->getId() == $document->getId()) {
+            if($site->getId() == $siteIdOfDocument) {
                 return $site;
             }
         }
 
         return null;
     }
 
+    public static function getSiteIdForDocument(Document $document): ?int
+    {
+        $pathSiteMappingCacheKey = 'path_site_mapping';
+
+        $pathMapping = Pimcore\Cache::load($pathSiteMappingCacheKey);
+
+        if (!$pathMapping) {
+            $pathMapping = [];
+        }
+
+        if (array_key_exists($document->getRealFullPath(), $pathMapping)) {
+            return $pathMapping[$document->getRealFullPath()];
+        }
+
+        foreach (self::getFullSiteListing() as $site) {
+            /** @var Site $site */
+            if (strpos($document->getRealFullPath(), $site->getRootPath() . '/') === 0 || $site->getRootDocument()->getId() == $document->getId()) {
+                $pathMapping[$document->getRealFullPath()] = $site->getId();
+                Pimcore\Cache::save($pathMapping, $pathSiteMappingCacheKey, ['system', 'resource'], null, 997);
+                return $site->getId();
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @return Site\Listing[]
+     */
+    private static function getFullSiteListing() : array
+    {
+        if (RuntimeCache::isRegistered(SELF::SITE_CACHE_KEY)) {
+            $sites = RuntimeCache::get(SELF::SITE_CACHE_KEY);
+        } else {
+            $sites = new Site\Listing();
```

Comment:
I'd create and cache the following array here: 
```php
[
            '/root/path/of/site-one' => 123,
            '/path/to/second/site' => 123,
        ]
```

We can then iterate over this array in `getSiteIdForDocument()` and check with `str_starts_with()` if the document is within a site. That should be quite quick and generic. 

Using the `RuntimeCache` shouldn't be necessary anymore then.

Reply:
Done.
 
Now we need to make a database call in `getSiteForDocument()` to get the site object, since we only have the `id` in the cache. Would it make sense to cache the site object as well? 

Or is it an overhead since the cache should be flushed every time the object is updated.

Reply:
@martineiber using `Site::getById()` is fine, it has a 2-level cache (runtime + core) build in 😉